pkg/collector/corechecks/snmp/snmp.go

```diff
@@ -0,0 +1,141 @@
+package snmp
+
+import (
+	"fmt"
+	"github.com/DataDog/datadog-agent/pkg/aggregator"
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/collector/check"
+	core "github.com/DataDog/datadog-agent/pkg/collector/corechecks"
+	"github.com/DataDog/datadog-agent/pkg/metrics"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"time"
+)
+
+const (
+	snmpCheckName = "snmp"
+)
+
+// Check aggregates metrics from one Check instance
+type Check struct {
+	core.CheckBase
+	config  snmpConfig
+	session sessionAPI
+	sender  metricSender
+}
+
+// Run executes the check
+func (c *Check) Run() error {
+	start := time.Now()
+
+	sender, err := aggregator.GetSender(c.ID())
+	if err != nil {
+		return err
+	}
+	c.sender = metricSender{sender: sender}
+
+	staticTags := c.config.getStaticTags()
+
+	var checkErr error
+	tags, checkErr := c.doRun(staticTags)
+	if checkErr != nil {
+		c.sender.serviceCheck("snmp.can_check", metrics.ServiceCheckCritical, "", tags, checkErr.Error())
+	} else {
+		c.sender.serviceCheck("snmp.can_check", metrics.ServiceCheckOK, "", tags, "")
+	}
+
+	c.sender.gauge("snmp.devices_monitored", float64(1), "", tags)
+
+	// SNMP Performance metrics
+	c.sender.monotonicCount("datadog.snmp.check_interval", float64(start.UnixNano())/1e9, "", tags)
+	c.sender.gauge("datadog.snmp.check_duration", float64(time.Since(start))/1e9, "", tags)
+	c.sender.gauge("datadog.snmp.submitted_metrics", float64(c.sender.submittedMetrics), "", tags)
+
+	// Commit
+	sender.Commit()
+	return checkErr
+}
+
+func (c *Check) doRun(staticTags []string) ([]string, error) {
+	tags := copyStrings(staticTags)
+
+	// Create connection
+	connErr := c.session.Connect()
+	if connErr != nil {
+		return tags, fmt.Errorf("snmp connection error: %s", connErr)
+	}
+	defer func() {
+		err := c.session.Close()
+		if err != nil {
+			log.Warnf("failed to close session: %v", err)
+		}
+	}()
+
+	// If no OIDs, try to detect profile using device sysobjectid
+	if !c.config.oidConfig.hasOids() {
+		sysObjectID, err := fetchSysObjectID(c.session)
+		if err != nil {
+			return tags, fmt.Errorf("failed to fetching sysobjectid: %s", err)
+		}
+		profile, err := getProfileForSysObjectID(c.config.profiles, sysObjectID)
+		if err != nil {
+			return tags, fmt.Errorf("failed to get profile sys object id for `%s`: %s", sysObjectID, err)
+		}
+		err = c.config.refreshWithProfile(profile)
+		if err != nil {
+			// Should not happen since the profile is one of those we matched in getProfileForSysObjectID
+			return tags, fmt.Errorf("failed to refresh with profile `%s` detected using sysObjectID `%s`: %s", profile, sysObjectID, err)
+		}
+	}
+	tags = append(tags, c.config.profileTags...)
+
+	// Fetch and report metrics
+	if c.config.oidConfig.hasOids() {
```

Comment:
do you need this check even if you successfully loaded a profile at line 74 ?

Reply:
We might have an empty profile, in that case, we can't do any fetch/report. That mimics what's done in Python: https://github.com/DataDog/integrations-core/blob/48f1b411c12e20f2edc1f065fe186065459781b0/snmp/datadog_checks/snmp/snmp.py#L419

It's debatable if we should or not send the uptime metric (see `c.config.addUptimeMetric()`) event if we don't get any metric from the profile 🤔